File: day6_orbitmap/python/day6_orbitmap_part_2.py
#!/bin/python

import sys
import io

def orbit_path(satellite, graph):
    p = satellite
    ret = []
    while True:
        p = graph.get(p)
        if p == None:
            return ret
        else:
            ret.append(p)


file_path = sys.argv[1]
graph = {}
with io.open(file_path, "r") as f:
    while True:
        line = f.readline()
        if line:
            tokens = line.split(")")
            satellite = tokens[1].strip()
            centre = tokens[0].strip()
            graph[satellite] = centre
        else:
            break

your_path = orbit_path("YOU", graph)
santa_path = orbit_path("SAN", graph)

total_transfers = 0
common_orbit = None
for p in your_path:
    common_orbit = p
    if p in santa_path:
        break
    total_transfers += 1

for p in santa_path:
    if p == common_orbit:
        break
    total_transfers += 1

print(total_transfers)

File: day6_orbitmap/rust/day6_orbitmap_part_2.rs
use std::env;
use std::fs::File;
use std::io::{BufRead, BufReader};
use std::collections::{HashMap, HashSet};

fn main() {
    let args: Vec<String> = env::args().collect();
    let file_path: &String = &args[1];

    let f = File::open(file_path).expect("Unable to open file");
    let f = BufReader::new(f);

    let mut planet_name_cache = HashSet<Rc<String>>::new();
    let mut graph = HashMap::new();
    for line in f.lines() {
        let line = line.expect("Unable to read line");
        let tokens:Vec<&str> = line.split(")").collect();
        let satellite = String::from(tokens[1]);
        let centre = String::from(tokens[0]);
        graph.insert(satellite, centre);
    }

    let you_path = get_path_to_galactic_centre("YOU", &graph);
    let santa_path = get_path_to_galactic_centre("SAN", &graph);
    let transfers = count_transfers(&you_path, &santa_path);

    println!("{}", transfers);
}


fn get_path_to_galactic_centre<'a>(satellite: &'a str, graph: &'a HashMap::<String, String>) -> Vec<&'a str> {
    let mut ret = Vec::new();
    let mut n: &str = satellite;
    loop {
        match graph.get(n) {
            Some(m) => {
                n = &m;
                ret.push(n);
            },
            None => return ret,
        }
    }
}


fn count_transfers(you_path: &Vec<&str>, santa_path: &Vec<&str>) -> usize {
    for (you_distance, &p) in you_path.iter().enumerate() {
        match santa_path.iter().position(|&s| s == p) {
            Some(santa_distance) => return you_distance + santa_distance,
            None => continue,
        }
    }
    panic!("No path");
}
